### ISSUE TITLE
Implement API ACL

### ISSUE THREAD

**Issue by danwdart** [dadi/publish#396]
Could use @dadi/api-wrapper through an isomorphic fetch.

**Comment by eduardoboucas** [comment]
Could you expand a bit on this one?

Not sure if this is what you're referring to, but we currently don't have server-side rendering for the UI components. It's something we discussed, but ultimately we felt that the potential performance gains weren't important enough to justify the added complexity.

We use API wrapper extensively in Publish to form any API queries ([`@dadi/api-wrapper-core`](https://www.npmjs.com/package/@dadi/api-wrapper-core), to be more precise, which is the non-Node version of API wrapper). My thinking was that instead of making the Publish server relay requests to API, the front-end could use `@dadi/api-wrapper-core` directly to get data from API, setting the [callback function](https://github.com/dadi/api-wrapper-core#defining-a-callback) to use [fetch](https://github.com/github/fetch) or any other Ajax library.

Does this align with what you're thinking?

**Comment by danwdart** [comment]
I understand the first point.
I see what you mean about the second point - so that we would just need to change the destination and add the auth headers - seeing as we're already using api-wrapper-core.

**Comment by danwdart** [comment]
Yes, I believe it's this one - https://github.com/dadi/publish/blob/master/frontend/lib/api-bridge-client.js - it should hit the API itself this time.

**Comment by danwdart** [comment]
This would probably require exposing an access token to the frontend code - if not a whole new pair of credentials - will try out this working before ACL

**Comment by eduardoboucas** [comment]
The idea with the new ACL piece is that each Publish user will have their own client ID+secret pair in API, so Publish will be able to request a bearer token on their behalf (by pinging the `/token` endpoint in API). The resulting bearer token will then be used by the front-end application to communicate with API directly.

**Comment by danwdart** [comment]
Yep awesome that's what I'm looking into atm

**Comment by eduardoboucas** [comment]
Bear in mind this doesn't exist in API yet. It will be released as 4.0 RC1 probably tomorrow, and I'm still writing the documentation as we speak. If you think it's easier to wait until the implementation is fully stable and documentation is available, it's totally fine to put this on hold and tackle something else in the meantime.

**Comment by danwdart** [comment]
I'm continuing on here - what I did is create a client ID and secret field in Users and use that to request documents, optionally waiting for a token to be stored before doing so.

**Comment by eduardoboucas** [comment]
Sorry, I'm not sure I understood that. In any case, I think it'd be useful if I shared what I had in mind for the Publish + API interaction with ACL in place.

When a user signs in to Publish, the details they entered will be sent to API (`/token` endpoint) to be exchanged for a bearer token. If it succeeds, the bearer token is stored somewhere in the client's machine (maybe a cookie, maybe local storage?).

As discussed, the Publish front-end app will communicate to API directly, instead of relaying requests through the back-end application. So, on every request Publish will take the bearer token that is stored in the client and append that to requests sent to API. Whatever comes back from API is what that client has access to, so there's nothing that Publish needs to do regarding permissions.

If a 401 comes back, it probably means the token has expired, so Publish needs to redirect the user to the sign in page again.

Does that make sense?

**Comment by danwdart** [comment]
I've made an initial commit for overriding the way we do most of that, which works for me at the moment, and it still needs 401 detection and saving of the token to localStorage.

**Comment by danwdart** [comment]
Looks also like the logout endpoint still needs to revoke the token too.
What's the token revoke endpoint? I'm looking at api docs now.

**Comment by danwdart** [comment]
The docs don't seem to specify but I'll try DELETE /token/{token}

**Comment by danwdart** [comment]
Okay, that doesn't exist, probably needs creating explicitly

**Comment by danwdart** [comment]
This RFC says we should probably add it to /revoke:
https://tools.ietf.org/html/rfc7009

**Comment by eduardoboucas** [comment]
Why does the logout endpoint need to revoke the token? Can't we simply remove it from the client? (Btw, we should probably use a cookie rather than localStorage, because localStorage can be weird, particularly when browsing in private mode)

**Comment by danwdart** [comment]
I think the difference between our approaches is that mine would fetch a new token every time the old one had expired. So that's why I didn't consider removing the token from the client. You're right, the client should keep one token as a "session", maybe later considering a "remember me" as fetching new tokens on demand.

So:

Login -> grab only one token and save to JS cookie, if 401, sorry your credentials weren't valid
If 200, use token for single session
When token expires / requests 401, redirect to login page
On logout, manually remove token and forget that it ever existed (but it still will do) - we get redirected to login page.

The only thing we probably want to be careful of with this approach is carefully 403ing where appropriate, to collections we're not allowed on, since 401ing will logout (which isn't correct)

I think I've got it all now - do you agree?

**Comment by eduardoboucas** [comment]
Exactly, that's what I'm thinking. I think for now it's okay to say that a session will only last as long as the bearer token does. Perhaps it would be nice if API allowed a configurable TTL per client, so you could have a longer TTL for a client that is expected to use their bearer token as a "session".

Your flow makes sense to me! 👍 

Re: 403 vs. 401, API should take care of that. Under normal circumstances, you shouldn't even see collections that you don't have access to – Publish uses the `/api/collections` endpoint to see what collections are available, and in 4.0 this list will reflect the client's permissions. However, if someone does try to access a collection they don't have access to, by manipulating the URL for example, API should respond with a 403 (or a 404, in some special cases). Either way, we can make Publish respond accordingly.

**Comment by danwdart** [comment]
Cool, will head on with this way of thinking this afternoon.
I will also see if I can do TTL stuff, but in a different PR.

**Comment by danwdart** [comment]
Integrating this into the Redux login stuff.
Not sure whether to set this cookie as a new Redux stored value as well, single point of truth and all that.

I think we go:

Login -> check user session with passport as usual, dispatch setRemoteUser, then fetch cookie.
Logout -> clear cookie

**Comment by danwdart** [comment]
Currently the above happens - but I think I'm catching in a couple of wrong places. It works, but it refreshes upon login and logout a little much. It still debugs a bunch, and I'm unsure of the Redux dispatch logic where I've touched it.
I've added a few more logs where we fail - I think that some of the @dadi/api-wrapper-core code is incorrect, as it does not reject some fetch requests such as config when they 401, it resolves them with undefined, which then I have to check for.
